src/coreclr/jit/fgprofilesynthesis.cpp

```diff
@@ -0,0 +1,917 @@
+            nextEdge->setLikelihood(0.0);
+        }
+
+        return;
+    }
+
+    // LOOP BACK EDGE heuristic
+    //
+    bool const isJumpEdgeBackEdge = IsLoopBackEdge(jumpEdge);
+    bool const isNextEdgeBackEdge = IsLoopBackEdge(nextEdge);
+
+    if (isJumpEdgeBackEdge != isNextEdgeBackEdge)
+    {
+        if (isJumpEdgeBackEdge)
+        {
+            JITDUMP(FMT_BB "->" FMT_BB " is loop back edge\n", block->bbNum, jump->bbNum);
+            jumpEdge->setLikelihood(0.9);
+            nextEdge->setLikelihood(0.1);
+        }
+        else
+        {
+            JITDUMP(FMT_BB "->" FMT_BB " is loop back edge\n", block->bbNum, next->bbNum);
+            jumpEdge->setLikelihood(0.1);
+            nextEdge->setLikelihood(0.9);
+        }
+
+        return;
+    }
+
+    // LOOP EXIT EDGE heuristic
+    //
+    // Consider: adjust probability if loop has multiple exit edges, so that
+    // overall exit probability is around 0.1.
+    //
+    bool const isJumpEdgeExitEdge = IsLoopExitEdge(jumpEdge);
+    bool const isNextEdgeExitEdge = IsLoopExitEdge(nextEdge);
+
+    if (isJumpEdgeExitEdge != isNextEdgeExitEdge)
+    {
+        if (isJumpEdgeExitEdge)
+        {
+            JITDUMP(FMT_BB "->" FMT_BB " is loop exit edge\n", block->bbNum, jump->bbNum);
+            jumpEdge->setLikelihood(0.1);
+            nextEdge->setLikelihood(0.9);
+        }
+        else
+        {
+            JITDUMP(FMT_BB "->" FMT_BB " is loop exit edge\n", block->bbNum, next->bbNum);
+            jumpEdge->setLikelihood(0.9);
+            nextEdge->setLikelihood(0.1);
+        }
+
+        return;
+    }
+
+    // RETURN heuristic
+    //
+    bool const isJumpReturn = (jump->bbJumpKind == BBJ_RETURN);
+    bool const isNextReturn = (next->bbJumpKind == BBJ_RETURN);
+
+    if (isJumpReturn != isNextReturn)
+    {
+        if (isJumpReturn)
+        {
+            jumpEdge->setLikelihood(0.2);
+            nextEdge->setLikelihood(0.8);
+        }
+        else
+        {
+            jumpEdge->setLikelihood(0.8);
+            nextEdge->setLikelihood(0.2);
+        }
+
+        return;
+    }
+
+    // IL OFFSET heuristic
+    //
+    // Give slight preference to bbNext
+    //
+    jumpEdge->setLikelihood(0.48);
+    nextEdge->setLikelihood(0.52);
+}
+
+//------------------------------------------------------------------------
+// AssignLikelihoodCond: update edge likelihood for a block that
+//   ends in a switch
+//
+// Arguments;
+//   block -- block in question (BBJ_SWITCH)
+//
+void ProfileSynthesis::AssignLikelihoodSwitch(BasicBlock* block)
+{
+    // Assume each switch case is equally probable
+    //
+    const unsigned n = block->NumSucc();
+    const weight_t p = 1 / (weight_t)n;
+
+    // Each unique edge gets some multiple of that basic probability
+    //
+    for (BasicBlock* const succ : block->Succs(m_comp))
+    {
+        FlowEdge* const edge = m_comp->fgGetPredForBlock(succ, block);
+        edge->setLikelihood(p * edge->getDupCount());
```

Comment:
Wouldn't all the heuristics in `AssignLikelihoodCond` apply? And need to get blended somehow?